src/Mvc/Mvc.Core/src/ModelBinding/Metadata/DefaultBindingMetadataProvider.cs

```diff
@@ -72,6 +73,73 @@ public void CreateBindingMetadata(BindingMetadataProviderContext context)
                 context.BindingMetadata.IsBindingAllowed = bindingBehavior.Behavior != BindingBehavior.Never;
                 context.BindingMetadata.IsBindingRequired = bindingBehavior.Behavior == BindingBehavior.Required;
             }
+
+            if (GetBoundConstructor(context.Key.ModelType) is ConstructorInfo constructorInfo)
+            {
+                context.BindingMetadata.BoundConstructor = constructorInfo;
+            }
+        }
+
+        internal static ConstructorInfo GetBoundConstructor(Type type)
+        {
+            if (type.IsAbstract || type.IsValueType || type.IsInterface)
+            {
+                return null;
+            }
+
+            var constructors = type.GetConstructors();
+            if (constructors.Length == 0)
+            {
+                return null;
+            }
+
+            return GetRecordTypeConstructor(type, constructors);
+        }
+
+        private static ConstructorInfo GetRecordTypeConstructor(Type type, ConstructorInfo[] constructors)
+        {
+            if (!IsRecordType(type))
+            {
+                return null;
+            }
+
+            // For record types, we will support binding and validating the primary constructor.
+            // There isn't metadata to identify a primary constructor. Our heuristic is:
+            // We require exactly one constructor to be defined on the type, and that every parameter on
+            // that constructor is mapped to a property with the same name and type.
+
+            if (constructors.Length > 1)
+            {
+                return null;
+            }
+
+            var constructor = constructors[0];
+
+            var parameters = constructor.GetParameters();
+            var properties = PropertyHelper.GetVisibleProperties(type);
+
+            for (var i = 0; i < parameters.Length; i++)
+            {
+                var parameter = parameters[i];
+                var mappedProperty = properties.FirstOrDefault(property =>
+                    string.Equals(property.Name, parameter.Name, StringComparison.Ordinal) &&
+                    property.Property.PropertyType == parameter.ParameterType);
+
+                if (mappedProperty is null)
+                {
+                    // No property found, this is not a primary constructor.
+                    return null;
+                }
+            }
+
+            return constructor;
+
+            static bool IsRecordType(Type type)
```

Comment:
for what it's worth, I think this check cuts out F# records from support in this feature. The other checks (single constructor, properties that match field names) seem to be met by F# records (though I'd want a test to confirm), but this detection is just un-circumventable.

Here's a [sharplab](https://sharplab.io/#v2:DYLgZgzgPg9gDgUwHYAIDKBPCAXBBbAWAChjsNEUAlBAYxQF4UBvFbACwEskBzEFHAE5duKAL5A=) that shows what F# records look like, they are detectable by attribute metadata.

Reply:
@baronfel \ @daniellittledev it looks like F# compiler generates attributes on the property rather than the parameter:

```
type Rec = { [<System.ComponentModel.BindableAttribute(true)>] thing: string }

[Bindable(true)]
[CompilationMapping(SourceConstructFlags.Field, 0)]
public string thing
{
    get
    {
        return thing@;
    }
}

public Rec(string thing)
{
    thing@ = thing;
}
```

Is there a gesture I'm missing that would cause it to produce these attributes on the parameters instead of the properties? In this PR, MVC looks for binding and validation specific metadata specifically on parameters.

Reply:
There's not a mechanism I'm aware of to influence how the F# compiler creates the constructor. @cartermp any ideas here?

Does this PR look for metadata on parameters because that's where the C# compiler emits them? If so there could be separate lookup paths for each one in lieu of some sort of unification work.

Reply:
> Does this PR look for metadata on parameters because that's where the C# compiler emits them

Yup.

>  If so there could be separate lookup paths for each one in lieu of some sort of unification work.

We'd initially wanted to support binding constructors in general, but don't feel confident about getting enough feedback for a correct implementation so close to a RC release. If we were to enable binding to ctors in a later release, F# records should work without having to do any additional work.

Reply:
Sounds good. I wanted to raise the question before this got committed to make sure that at least the concept of supporting F# records was discussed. Thanks for engaging :)

Reply:
My take on this is that we should hold off on considering F# records in scope for now. My money is on there needing to be further changes before binding to C# records is considered 100% done and stable